### PR TITLE
Fix IPv6 compression

### PR DESCRIPTION
**Pull request by apmoriarty**
Previous algorithm was incorrectly replacing a single leading hextet of all zeroes with `::` which led to invalid IPv6 addresses that would throw an exception when attempting to parse.

### CODE REVIEW

File: core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java [hunk 1]
```diff
@@ -98,48 +105,87 @@ public static IpV6Address parse(String address) {
...
-            int count = 0;
-            for (int i = 0; i < 8; i++) {
-                if (address[i] == 0) {
-                    count++;
-                } else {
-                    if (count > length) {
-                        startSkip = i - count;
-                        length = count;
+            for (int i = 0; i < hextets.size(); i++) {
+                if (isHextetAllZeroes(hextets.get(i))) {
```
Comment (ivakegg) [inline core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java]: could this be simply (address[i] == 0)

File: core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java [hunk 2]
```diff
@@ -98,48 +105,87 @@ public static IpV6Address parse(String address) {
...
+                            len++;
+                        } else {
+                            break;
+                        }
+                    }
+
+                    if (len >= 2 && len > zeroHextetLength) {
+                        zeroHextetStart = i;
+                        zeroHextetLength = len;
+                        log.trace("found sequence at index: {} length: {}", zeroHextetStart, zeroHextetLength);
```
Comment (ivakegg) [inline core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java:144]: could advance i here to avoid checking already skipped zeros

File: core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java [hunk 3]
```diff
@@ -98,48 +105,87 @@ public static IpV6Address parse(String address) {
...
-                    count++;
-                } else {
-                    if (count > length) {
-                        startSkip = i - count;
-                        length = count;
+            for (int i = 0; i < hextets.size(); i++) {
+                if (isHextetAllZeroes(hextets.get(i))) {
+                    int len = 1;
+                    for (int j = i + 1; j < hextets.size(); j++) {
+                        if (isHextetAllZeroes(hextets.get(j))) {
```
Comment (ivakegg) [inline core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java]: could be address[j] == 0

File: core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java [hunk 4]
```diff
@@ -126,10 +126,10 @@ public static String toString(short[] address, boolean zeroPadded, boolean skipZ
         int zeroHextetLength = -1;
         if (skipZeros) {
             for (int i = 0; i < hextets.size(); i++) {
-                if (isHextetAllZeroes(hextets.get(i))) {
+                if (hextets.get(i).equals("0")) {
```
Comment (ivakegg) [inline core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java:131]: if zeropadding, then hextets.get(i).equals("0000")

Comment (ivakegg) [reply core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java:131]: perhaps we need some test with zero passing on

Comment (apmoriarty) [reply core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java:131]: per discussion, verified the method is always called with different flags set so this case should never be hit. Added a precondition check to fail if the method is ever erroneously called with true/true or false/false.

**Review by ivakegg — APPROVED** [review]

**Review by hlgp — APPROVED** [review]

File: core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java [hunk 5]
```diff
@@ -98,48 +106,80 @@ public static IpV6Address parse(String address) {
     }
 
     public static String toString(short[] address, boolean zeroPadded, boolean skipZeros) {
-        StringBuilder builder = new StringBuilder(39);
-        int startSkip = -1;
-        int length = 0;
+
+        Preconditions.checkArgument(zeroPadded != skipZeros, "cannot zero pad address and skip zeros");
```
Comment (alerman) [inline core/utils/type-utils/src/main/java/datawave/data/type/util/IpV6Address.java:110]: This would also fail if both not zero padded and not skip zeros. Intended?
